### PR TITLE
Fixes #29959 - Support yum migration

### PR DESCRIPTION
**Pull request by jlsherrill**

### CODE REVIEW

**Comment by jlsherrill** [comment]
Not ready for a review yet

**Comment by ianballou** [comment]
Looks like the same thing happens if you remove steps 4,5, and 6.

**Comment by jlsherrill** [comment]
@ianballou ohhh actually for errata, 'migrated_pulp3_href' isn't supposed to be set on any errata.  Only the 'erratum_pulp3_href' within RepositoryErratum should be.  

The reason for this is that errata in pulp3 are somewhat unique per repo.  We 'combine' them all into one object in katello.  
If you're seeing any errors as a result of this, let me know!

**Comment by ianballou** [comment]
> @ianballou ohhh actually for errata, 'migrated_pulp3_href' isn't supposed to be set on any errata. Only the 'erratum_pulp3_href' within RepositoryErratum should be.
> 
> The reason for this is that errata in pulp3 are somewhat unique per repo. We 'combine' them all into one object in katello.
> If you're seeing any errors as a result of this, let me know!

That makes sense! I am seeing an error, the content switchover is not liking the nil `migrated_pulp3_href` for errata.

File: app/services/katello/pulp3/migration.rb [hunk 1]
```diff
@@ -120,18 +163,52 @@ def process_distributions(pulp3_api, dist_hrefs_list)
...
         end
-        unmigrated_units.select(:id, :pulp_id).find_in_batches(batch_size: GET_QUERY_ID_LENGTH) do |needing_hrefs|
+        map
+      end
+
+      def import_errata
+        to_import = []
+        repo_version_map = generate_repo_version_map
+        Katello::Erratum.select(:id, :pulp_id).find_in_batches(batch_size: GET_QUERY_ID_LENGTH) do |needing_hrefs|
           migrated_units = pulp2_content_api.list(pulp2_id__in: needing_hrefs.map { |unit| unit.pulp_id }.join(','))
```
[the remark this answers is not in the document]
Comment (jlsherrill) [reply app/services/katello/pulp3/migration.rb:179]: ahh, i will fix that!

**Comment by jlsherrill** [comment]
@ianballou updated to now import all RepositoryErratum href's

**Comment by jlsherrill** [comment]
Thanks @ianballou ! Fixed!

**Comment by ianballou** [comment]
> I did another test where I synced 3 repos (needed_errata, rhel 8 os, rhel 8 appstream), migrated, stuck the 3 in a content view, and migrated again. Pre-content view, the Repository Errata all got their Pulp 3 hrefs. Afterwards, though, there were missing Pulp 3 hrefs. I'm going to try again in case it was a fluke.

I tested this again and content-view Repository Errata are still not receiving their Pulp 3 hrefs, so it seems it's a real issue.

**Comment by ianballou** [comment]
Also there is a success message after this error, think we need a failure message.

**Comment by ianballou** [comment]
If I don't publish my content view twice and avoid the cardinality error above, I am able to do the post migration check and the content switch-over.  I'm not sure if it's related, but `ContentGuard` refresh is getting a 404 and causing Pulp3 repo syncs to fail now, however.

For my AppStream repo I got this error after commenting out the `ContentGuard` stuff:
```
NoMethodError: undefined method `pulp_href' for nil:NilClass
Backtrace:
/home/vagrant/katello/app/services/katello/pulp3/repository.rb:305:in `secure_distribution_options'
/home/vagrant/katello/app/services/katello/pulp3/repository.rb:233:in `update_distribution'
/home/vagrant/katello/app/services/katello/pulp3/repository.rb:216:in `refresh_distributions'
/home/vagrant/katello/app/lib/actions/pulp3/repository/refresh_distribution.rb:21:in `invoke_external_task'
```

I know some stuff isn't working right now due to Pulp 3 bugs, is this part of that?

On my BaseOS repo I'm also hitting this Pulp error during `pulp_rpm.app.tasks.synchronizing.synchronize`:

```
Katello::Errors::Pulp3Error: No declared artifact with relative path "repodata/adff8e65-8c3c-4ffb-bb72-5d4116fab7c9" for content "<RepoMetadataFile: pk=0c7e0440-aed2-44f0-9053-dc1d0a899330>"
```

**Comment by ianballou** [comment]
Since a lot of migration things are being fixed here, here's a Docker issue I found:
1) Sync any docker repo and add it to a content view
2) Make a filter and publish
3) Migrate
4) Post migration check
5) See this: `ERROR: Non-archived docker repositories with ID [13,14] have no distribution reference`

**Review by ianballou — COMMENTED** [review]
I hit an issue during the content switchover where two Pulp 2 package groups were exactly the same and were migrated into one Pulp 3 package group.  This caused a uniqueness constraint violation:
```
ActiveRecord::RecordNotUnique: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_katello_package_groups_on_pulp_id"
DETAIL:  Key (pulp_id)=(/pulp/api/v3/content/rpm/packagegroups/bd0766fe-d89a-40be-aafb-ebd670f3e9b2/) already exists.
```

The repositories were the following:
http://mirror.centos.org/centos/8/AppStream/x86_64/os
http://yum.oracle.com/repo/OracleLinux/OL8/appstream/x86_64/

The package group was the following:
"Graphical Administration Tools"

**Comment by ianballou** [comment]
I hit another issue.  After doing a successful content switchover, I tried to re-publish my content view.  I got a 404 on a package group.  After some investigation, it looks like the Pulp 3 hrefs are out of date in the Katello DB.  The package group existed, but with a different href.  Is Pulp 3 re-creating the IDs during a re-migration?

**Comment by jlsherrill** [comment]
@ianballou i know you may not get this, but did you do a pulp3 db reset in between migrations when you saw dissappearing rpm units?

**Comment by ianballou** [comment]
I didn't do any Pulp 3 DB resets.

**Comment by ianballou** [comment]
I was not able to reproduce the content unit 404 error I got above.  I migrated between CV publishes and deleted a CVV too.

**Comment by ianballou** [comment]
> I hit an issue during the content switchover where two Pulp 2 package groups were exactly the same and were migrated into one Pulp 3 package group. This caused a uniqueness constraint violation:

Is ^ ready for re-testing?

**Comment by jlsherrill** [comment]
Yes, support was added here: https://github.com/Katello/katello/pull/8737/files#diff-dad89bab693431f51527ba7a8b266671R70

**Comment by ianballou** [comment]
Seeing a new issue during switchover:

```
NameError: uninitialized constant Katello::RepositoryYumMetadataFile
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/active_support.rb:80:in `block in load_missing_constant'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/active_support.rb:9:in `without_bootsnap_cache'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/active_support.rb:80:in `rescue in load_missing_constant'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/active_support.rb:59:in `load_missing_constant'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/inflector/methods.rb:284:in `const_get'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/inflector/methods.rb:284:in `block in constantize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/inflector/methods.rb:280:in `each'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/inflector/methods.rb:280:in `inject'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/inflector/methods.rb:280:in `constantize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/core_ext/string/inflections.rb:68:in `constantize'
/home/vagrant/katello/app/models/katello/concerns/pulp_database_unit.rb:39:in `repository_association_class'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:76:in `block (2 levels) in combine_duplicate_content_types'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:73:in `each'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:73:in `block in combine_duplicate_content_types'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:71:in `each'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:71:in `combine_duplicate_content_types'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:21:in `block (2 levels) in run'
/home/vagrant/katello/app/lib/katello/logging.rb:6:in `time'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:21:in `block in run'
/home/vagrant/katello/app/lib/katello/logging.rb:6:in `time'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:17:in `run'
/home/vagrant/katello/lib/katello/tasks/pulp3_content_switchover.rake:12:in `block (3 levels) in <top (required)>'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.3/lib/active_record/connection_adapters/abstract/database_statements.rb:280:in `block in transaction'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.3/lib/active_record/connection_adapters/abstract/transaction.rb:280:in `block in within_new_transaction'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.3/lib/active_record/connection_adapters/abstract/transaction.rb:278:in `within_new_transaction'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.3/lib/active_record/connection_adapters/abstract/database_statements.rb:280:in `transaction'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.3/lib/active_record/transactions.rb:212:in `transaction'
/home/vagrant/katello/lib/katello/tasks/pulp3_content_switchover.rake:10:in `block (2 levels) in <top (required)>'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli/exec.rb:75:in `load'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli/exec.rb:75:in `kernel_load'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli/exec.rb:28:in `run'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli.rb:424:in `exec'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli.rb:27:in `dispatch'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli.rb:18:in `start'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/exe/bundle:30:in `block in <top (required)>'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/exe/bundle:22:in `<top (required)>'
/opt/rh/rh-ruby25/root/usr/bin/bundle:23:in `load'
/opt/rh/rh-ruby25/root/usr/bin/bundle:23:in `<main>'

Caused by:
NameError: uninitialized constant Katello::RepositoryYumMetadataFile
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/active_support.rb:61:in `block in load_missing_constant'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/active_support.rb:17:in `allow_bootsnap_retry'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/active_support.rb:60:in `load_missing_constant'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/inflector/methods.rb:284:in `const_get'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/inflector/methods.rb:284:in `block in constantize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/inflector/methods.rb:280:in `each'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/inflector/methods.rb:280:in `inject'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/inflector/methods.rb:280:in `constantize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/core_ext/string/inflections.rb:68:in `constantize'
/home/vagrant/katello/app/models/katello/concerns/pulp_database_unit.rb:39:in `repository_association_class'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:76:in `block (2 levels) in combine_duplicate_content_types'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:73:in `each'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:73:in `block in combine_duplicate_content_types'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:71:in `each'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:71:in `combine_duplicate_content_types'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:21:in `block (2 levels) in run'
/home/vagrant/katello/app/lib/katello/logging.rb:6:in `time'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:21:in `block in run'
/home/vagrant/katello/app/lib/katello/logging.rb:6:in `time'
/home/vagrant/katello/app/services/katello/pulp3/migration_switchover.rb:17:in `run'
/home/vagrant/katello/lib/katello/tasks/pulp3_content_switchover.rake:12:in `block (3 levels) in <top (required)>'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.3/lib/active_record/connection_adapters/abstract/database_statements.rb:280:in `block in transaction'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.3/lib/active_record/connection_adapters/abstract/transaction.rb:280:in `block in within_new_transaction'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activesupport-6.0.3.3/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.3/lib/active_record/connection_adapters/abstract/transaction.rb:278:in `within_new_transaction'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.3/lib/active_record/connection_adapters/abstract/database_statements.rb:280:in `transaction'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.3/lib/active_record/transactions.rb:212:in `transaction'
/home/vagrant/katello/lib/katello/tasks/pulp3_content_switchover.rake:10:in `block (2 levels) in <top (required)>'
/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli/exec.rb:75:in `load'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli/exec.rb:75:in `kernel_load'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli/exec.rb:28:in `run'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli.rb:424:in `exec'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli.rb:27:in `dispatch'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/cli.rb:18:in `start'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/exe/bundle:30:in `block in <top (required)>'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/bundler-1.16.1/exe/bundle:22:in `<top (required)>'
/opt/rh/rh-ruby25/root/usr/bin/bundle:23:in `load'
/opt/rh/rh-ruby25/root/usr/bin/bundle:23:in `<main>'
Tasks: TOP => katello:pulp3_content_switchover
(See full trace by running task with --trace)
```

File: app/services/katello/pulp3/migration.rb [hunk 2]
```diff
@@ -47,43 +54,92 @@ def pulp2_repositories_api
...
+        [YumMetadataFile]
       end
 
       def content_types_for_migration
         content_types = @repository_types.collect do |repository_type_label|
           Katello::RepositoryTypeManager.repository_types[repository_type_label].content_types_to_index
         end
 
-        content_types.flatten
+        content_types.flatten - self.ignorable_content_types
```
Comment (ianballou) [inline app/services/katello/pulp3/migration.rb:70]: `NoMethodError: undefined method 'ignorable_content_types' for #<Katello::Pulp3::Migration:0x0000000011f82f80>`

**Review by ianballou — COMMENTED** [review]
Hitting an issue during switchover

File: app/services/katello/pulp3/migration.rb (hunk 2 above)
Comment (ianballou) [reply app/services/katello/pulp3/migration.rb:70]: Works otherwise though if I just hard code in the ignorable content types

**Comment by jlsherrill** [comment]
Yeah sorry, I caught that after I pushed and didn't have time to fix, will
do that

**Review by ianballou — COMMENTED** [review]
I think this is ready to merge as soon as the tests pass.

**Review by ianballou — APPROVED** [review]

**Pull request merged**